testkit/jvm/src/main/scala/zio/test/ZIOTest.scala

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright 2019 John A. De Goes and the ZIO Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package zio.test
+
+import zio._
+import zio.internal.{ Platform, PlatformLive }
+
+abstract class ZIOTest[R, E] { self =>
+
+  /**
+   * Construct your environment here.
+   */
+  val environment: Managed[Nothing, R]
```

Comment:
This assumes that tests create their own environment, which feels too concrete (i.e. doesn't fit our usecase!)
Also, `final def main` removes control over running an individual test from the user, pinning a single strategy...
What do you think about `ZTest` referencing the `Runner` that it want to run on and the `Runner` providing an implementation of `main` instead ?

Reply:
Here's the problem I'm trying to solve:

 - I want to be able to run a suite (existing inside a class) by running its main function

Everything else falls out of here.

There should be other ways, but you can see why we need the environment and other details inside this class: because otherwise we cannot provide runnability on subclasses.

Now, we can give up on runnability. But I think it's quite nice and most IDEs can already run files with a click or two. So it's out of the box ease-of-use.

We can tweak this design, but one of the important details I've been punting on is the extent to which the user must compose specs, and the extent to which the library does that using reflection.

If the user composes everything, it will be more verbose and painful for common cases, but also you can achieve other benefits, such as keeping `R` and so forth polymorphic and pushing that up higher and higher. OTOH with reflection, you need some unit at which everything has been packaged together in a form that's ready to run.

Reply:
That's fine – delegating `main` implementation to someone else doesn't hurt runnability – we can have a default implementor.

Reply:
I will pull out logic of `main` into something better, and it can be used by `main` and something else. I still think it should be separate from `runner` since that class should be dead simple, going through the tree, running each thing, and collecting the outputs (either sequentially or in parallel).

Reply:
Thanks! 👍